drv/mb85rsxx-fram/src/lib.rs

```diff
@@ -0,0 +1,489 @@
+#[must_use = "a Fram does nothing if constructed but not read from or written to"]
+pub struct Fram<S: SpiServer, const ID: u16> {
+    spi: SpiDevice<S>,
+}
+
+/// A generic Fujitsu FRAM chip with its write enable latch set.
+///
+/// This type is returned by [`Fram::write_enable`], and will unset the write
+/// latch when it's dropped. This way, the FRAM remains in the write-protected
+/// state when you're not actively trying to write to it.
+///
+/// To write to the FRAM chip, use [`WritableFram::write`]. This type also
+/// exposes a [`WritableFram::read`] method, so the FRAM chip may also be read
+/// from while the write enable latch is set.
+#[must_use = "a WritableFram does nothing if constructed but not read from \
+    or written to"]
+pub struct WritableFram<'fram, S: SpiServer, const ID: u16>(
+    &'fram Fram<S, { ID }>,
+);
+
+#[derive(Copy, Clone, PartialEq, Eq)]
+pub struct FramId {
+    pub mfg_id: u8,
+    pub product_id: u16,
+}
+
+#[derive(Eq, PartialEq, Copy, Clone, counters::Count)]
+pub enum FramInitError {
+    SpiError(#[count(children)] SpiError),
+    UnknownManufacturerId(u8),
+    UnexpectedProductId { expected: u16, actual: u16 },
+}
+
+/// Errors returned by the [`Fram::read`] and [`WritableFram::write`] methods.
+#[derive(Eq, PartialEq, Copy, Clone, counters::Count)]
+pub enum FramError {
+    SpiError(#[count(children)] SpiError),
+    InvalidAddr,
+    /// The write is longer than the highest address, would wrap around to the
+    /// beginning of the FRAM!
+    ///
+    /// You probably don't want that.
+    WouldWrap,
+}
+
+#[derive(Eq, PartialEq, Copy, Clone, counters::Count)]
+// We don't actually use all of these, because this driver doesn't currently
+// use all the FRAM chip's features (e.g. bank protection, sleep mode, etc).
+// But, I wanted to write out all supported opcodes anyway.
+#[allow(dead_code)]
+#[repr(u8)]
+enum Opcode {
+    /// Set the write enable latch (WREN)
+    SetWriteEn = 0b0000_0110,
+    /// Reset the write enable latch (WRDI)
+    ResetWriteEn = 0b0000_0100,
+    /// Read the status register (RDSR)
+    ReadStatus = 0b0000_0101,
+    /// Write to the status register (WRSR)
+    WriteStatus = 0b0000_0001,
+    /// Read from memory (READ)
+    Read = 0b0000_0011,
+    /// Write to memory (WRITE)
+    Write = 0b0000_0010,
+    /// Read Device ID (RDID)
+    ReadId = 0b1001_1111,
+    /// Sleep mode (SLEEP)
+    Sleep = 0b1011_1001,
+    /// Reserved for future use (RFU).
+    ///
+    /// Probably don't send this command. You may regret this.
+    Reserved = 0b0000_1011,
+}
+
+bitflags! {
+    #[derive(Copy, Clone, PartialEq, Eq)]
+    pub struct Status: u8 {
+        /// Write enable latch
+        const WEL = 1 << 1;
+        /// Block protect 0
+        const BP0 = 1 << 2;
+        /// Block protect 1
+        const BP1 = 1 << 3;
+        /// Status register write protect enabled
+        const WPEN = 1 << 7;
+    }
+}
+
+#[derive(Eq, PartialEq, Copy, Clone, counters::Count)]
```

Comment:
So, hmm. I found it quite useful for debugging the driver implementation, but I agree that it may not be that useful for debugging code depending on this. I do, however, think it's quite nice to have separate ringbuf events for "starting a read/write" and "did a read/write", so that there's a way to see if the FRAM driver is stuck while waiting for a read/write operation to complete (which is why the wrappers exist). But, higher-level code can do that.

Reply:
(i've since removed the ringbuf)